### PR TITLE
feat(e2e): add support for puppteer v19

### PR DESCRIPTION
**Pull request by rwaskiewicz**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [ ] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been reviewed and added / updated if needed (for bug fixes / features) - https://github.com/ionic-team/stencil-site/pull/932
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Unit tests (`npm test`) were run locally and passed
- [x] E2E Tests (`npm run test.karma.prod`) were run locally and passed
- [x] Prettier (`npm run prettier`) was run locally and passed

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

GitHub Issue Number: N/A

## What is the new behavior?

this pull request adds support for puppeteer v11 through v19.

each commit is described below:
<details><summary>Puppeteer v11, v12, v13</summary>

v11, v12, v13 all did not require any changes to the stencil codebase.
</details>

<details><summary>Puppeteer v14</summary>

the package is now declared in `package.json` using a caret ("^")
instead of a tilde ("~"). the author (rwaskiewicz) was being cautious
when making the change to support v10 of stencil in
https://github.com/ionic-team/stencil/pull/2934, to a degree that is no
longer necessary

v14 of puppeteer introduced private fields in their source code. this
affected stencil's validation scripts for type declaration files, as
the default settings would create a TypeScript `program` that could not
be properly compiled/validated.

the `moduleResolution` and `target` fields needed to be explictly
set to allow the transpiled `.d.ts` to pass validation. this is a result
of cascading defaults where for the `createProgram` call:
- the `target` field defaults to "ES3"
- the `target` default causes the `module` field to default to
  "CommonJS"
- the `module` default causes `moduleResolution` to default to
  "classic", which can't resolve the new type declaration file. as a
  result, we explicitly set this field
- the `target` field is set to support the private identifiers in
  puppeteer
</details>

<details><summary>Puppeteer v15</summary>

this commit increments the supported version of puppeteer from v14 to
v15.

starting with puppeteer v15, the library performs type
inference/deduction for the `evaluate()` function. this commit updates
the types (often removing them) at the advice spelled out in
https://github.com/puppeteer/puppeteer/pull/8547.
</details>

<details><summary>Puppeteer v16</summary>

v16 did not require any changes in the codebase, other than incrementing
the list of supported versions of the library.
</details>

<details><summary>Puppeteer v17</summary>
this commit increments the supported version of puppeteer from v16 to
v17.

puppeteer v17 made accessing a puppeteer `ExecutionContext`, an entity
used to run javascript, an internal entity. previously, stencil would
directly access retrieve an `ExecutionContext` from a helper function
that is no longer exposed. to work around this in puppeteer v17+, two
different strategies for getting access to this entity have been added.
Each strategy is dependent on how we were previously accessing the
`ExecutionContext`.

1. `ElementHandle` Scenario

in this scenario, an `ExecutionContext` was being pulled off an
`ElementHandle` instance.

the suggested way of getting an `ExecutionContext` in puppeteer v17+ for
an `ElementHandle` is through the `frame` getter on the `ElementHandle`
instance. doing so does not work in puppeteer v16 and below. for those
versions of puppeteer v16 and below, stencil will default to the original
`executionContext()` behavior. otherwise, the return value of the
`frame` getter is returned from a new utility method.

in order to determine which version of puppeteer is used, a mechanism
for reading the puppeteer `package.json#version` field has been added.
this avoids clunky prototype lookups that have been used in the past,
and are not always the safest way to detect the version of puppeteer
being used (e.g. a field may exist on the prototype chain of an object
in 2 different versions of puppeteer, but do very different things).

2. `JSHandle` Scenario

accessing a `JSHandle`'s `ExecutionContext` is necessary in
`puppeteer-event.ts`. because this is the only instance where stencil
would get an `ExecutionContext` from a `JSHandle`, no utility function
for retrieving an `ExecutionContext` was created. rather, the same
effect can be achieved in a backwards compatible way by calling
`evaluate()` directly on the `JSHandle` instance.

we do not call `.asElement()` on the `JSHandle` instance and
subsequently use the "`ElementHandle` Scenario" described above as a
`JSHandle` does not always have an element associated with it, making it
impossible to get an `ExecutionContext` in such instances
</details>

<details><summary>Puppeteer v18</summary>

puppeteer v18 did not include any breaking changes that required major
breaking changes to stencil
</details>

<details><summary>Puppeteer v19</summary>

puppeteer v19 did not include any breaking changes that required major
breaking changes to stencil
</details>

<details><summary>Node, NPM versions</summary>

this pr increments the versions of node & npm used to run our karma tests.
the newest version of puppeteer requires use to upgrade these. consumers
of stencil and puppeteer should refer to the puppeteer breaking changes
guides (owned by the puppeteer team) to verify they are using the correct
minimum versions of node/npm
</details>

## Does this introduce a breaking change?

- [x] Yes
- [ ] No

## Testing

This branch will be used for all manual testing.

### Step 0 - Verify Type Checking on Stencil
Create `test-pptr-w-tsc.sh` in the root of the Stencil repo with the following contents:
```sh
#! /bin/bash
set -e

for VERSION in {10..19}
do
    npm i puppeteer@"${VERSION}"
    npm ci
    npm run clean
    npm run build
done
```

Run the script `./test-pptr-w-tsc.sh`. This will verify that _if_ we need to roll back the version of puppeteer, the code type checks in a backwards compatible way. Note, this step _only_ checks the type checking, not the impl details of puppeteer and how they integrate with Stencil

### Step 1 - Build this Branch
After completing Step 0, be sure to `git reset --hard` to restore anything that changed

To build it, checkout this branch and run the following:
```sh
npm ci && \
npm run clean && \
npm run build && \
npm pack
```
Note the path of the created tarball. We'll use that and refer to it as `TARBALL_PATH`

### Step 2 - Smoke Test `create-stencil` Output

Create a small Stencil component library:
```sh
cd /tmp && \
npm init stencil component pptr && \
cd pptr && \
npm i && \
git init && \
git add -A && \
git commit -m 'init with cli'
```

In the root of the `pptr/` dir, create `test-version.sh` with the following:
```sh
#! /bin/bash
set -e

if [ -z "$1" ]; then
    echo "Please provide the path to the tarball for Stencil"
    exit 1
fi

npm i "$1"
for VERSION in {10..19}
do
    npm i puppeteer@"${VERSION}"
    npm run test
done
```

Finally, run `./test-version.sh <TARBALL_PATH>`. This will exercise the output of the `create-stencil` cli against the latest release for every major version of Puppeteer, starting with v10 and ending in v19. For all 10 versions, you should see something like:
```
> pptr@0.0.1 test
> stencil test --spec --e2e

[09:03.7]  @stencil/core
[09:03.9]  [LOCAL DEV] 🔋
[09:03.9]  testing e2e and spec files
[09:04.8]  build, pptr, dev mode, started ...
[09:04.8]  transpile started ...
[09:07.8]  transpile finished in 2.95 s
[09:07.8]  copy started ...
[09:07.8]  generate custom elements started ...
[09:07.8]  generate lazy started ...
[09:07.9]  copy finished (0 files) in 114 ms
[09:08.1]  generate custom elements finished in 317 ms
[09:08.1]  generate lazy finished in 316 ms
[09:08.1]  build finished in 3.36 s

[09:08.1]  jest args: --spec --e2e --max-workers=8
 PASS  src/utils/utils.spec.ts
 PASS  src/components/my-component/my-component.spec.ts
 PASS  src/components/my-component/my-component.e2e.ts

Test Suites: 3 passed, 3 total
Tests:       8 passed, 8 total
Snapshots:   0 total
Time:        1.673 s, estimated 2 s
Ran all test suites.
```

### Step 3 - Test this Branch as a Drop-In Replacement for Stencil in Ionic Framework
Check out the Ionic Framework repository, and install this branch of Stencil in core. 
```sh
cd core/ && \
npm i <PATH_TO_TARBALL>
```
We want to test that Puppeteer v10 works in the Ionic Framework with our changes from this branch applied. To do so, run the e2e tests:
```sh
npm run test.e2e
```

if everything passes, please move on to Step 4. It uses the work we just did.

### Step 4 - This this Branch with Puppeteer v10->v19 in Ionic Framework
Before starting this step, complete all instructions for Step 3

In this step, we want to use the Ionic Framework as an example of a mature codebase using the E2E testing suite. We'll use a script to install each version of Puppeteer and run the test suite against our Stencil build and a specific version of Puppeteer.

To automate this process, create `test-pptr-framework.sh` in the `core/` directory of the Ionic Framework repo you have:
```sh
#! /bin/bash
set -e

for VERSION in {10..19}
do
    npm i puppeteer@"${VERSION}"
    npm run test.e2e
done
```
And run it with `./test-pptr-framework.sh`. The tests _should pass_ for each version.

However! Please note:
- This is a long running script 😅 
- I saw flakiness in v16 & v17 on the Ionic Framework. This appears to be intermittent, and resolved in v18, v19 

### Why Does this PR Target Stencil v3? Why Not Add this to v2?
[Puppeteer v14 dropped support for Node 12](https://github.com/puppeteer/puppeteer/releases/tag/v14.0.0), which Stencil v2 still supports. We still have users on Node 12 (as well as our CI pipeline testing against it for them), making it impossible to merge this without forcing it (and subsequent PRs) or removing Node 12 checks from the pipeline.

### CODE REVIEW

File: src/sys/node/node-sys.ts [hunk 1]
```diff
@@ -607,8 +607,8 @@ export function createNodeSys(c: { process?: any } = {}): CompilerSystem {
     '@types/jest': { minVersion: '24.9.1', recommendedVersion: '27.0.3', maxVersion: '27.0.0' },
     jest: { minVersion: '24.9.1', recommendedVersion: '27.0.3', maxVersion: '27.0.0' },
     'jest-cli': { minVersion: '24.9.0', recommendedVersion: '27.4.5', maxVersion: '27.0.0' },
-    puppeteer: { minVersion: '10.0.0', recommendedVersion: '13.5.2' },
-    'puppeteer-core': { minVersion: '10.0.0', recommendedVersion: '13.5.2' },
+    puppeteer: { minVersion: '1.19.0', recommendedVersion: '19' },
+    'puppeteer-core': { minVersion: '1.19.0', recommendedVersion: '19' },
```
Comment (tanner-reits) [inline src/sys/node/node-sys.ts:611]: `1.19.0` doesn't seem right so just calling out to be safe

**Review by tanner-reits — CHANGES_REQUESTED** [review]
Just one small callout I wanted to verify

File: src/sys/node/node-sys.ts (hunk 1 above)
Comment (rwaskiewicz) [reply src/sys/node/node-sys.ts]: Nope - bad merge conflict resolution. Fixed in 0e4cb95cb

**Review by tanner-reits — APPROVED** [review]

**Review by alicewriteswrongs — CHANGES_REQUESTED** [review]
this all looks straightforward enough, I ran through the testing steps as well and confirmed that everything looks good. Just had one question about the version range.

File: src/sys/node/node-sys.ts [hunk 2]
```diff
@@ -607,8 +607,8 @@ export function createNodeSys(c: { process?: any } = {}): CompilerSystem {
     '@types/jest': { minVersion: '24.9.1', recommendedVersion: '27.0.3', maxVersion: '27.0.0' },
     jest: { minVersion: '24.9.1', recommendedVersion: '27.0.3', maxVersion: '27.0.0' },
     'jest-cli': { minVersion: '24.9.0', recommendedVersion: '27.4.5', maxVersion: '27.0.0' },
-    puppeteer: { minVersion: '10.0.0', recommendedVersion: '13.5.2' },
```
[the remark this answers is not in the document]
Comment (rwaskiewicz) [reply src/sys/node/node-sys.ts:610]: 👍 done in https://github.com/ionic-team/stencil/pull/3810/commits/303e7a71faabc422b606d1257e6b106fd3408a81

**Review by alicewriteswrongs — APPROVED** [review]

**Comment by rwaskiewicz** [comment]
@alicewriteswrongs it appears that when we add `maxVersion` to our requirements for dependencies, that breaks our CI workflow for projects that don't have/need puppeteer installed. I've reverted that change as a result and created STENCIL-650 to look into this further

**Review by alicewriteswrongs — APPROVED** [review]
lgtm!

**Pull request merged**